### ISSUE TITLE
[v2.0.0] Query exec API

### ISSUE THREAD

**Issue by AlexPikalov** [AlexPikalov/cdrs#189]
There are some thought about how could Query Execution API could look like in version 2 of CDRS. Any discussion and proposals are very welcome.

According to Cassandra binary protocol specification and existing CDRS types system each query has to have such fields that basically could be represented by following struct:

```rust
pub struct Query {
    pub query: String,

    pub consistency: Consistency,
    pub values: Option<Vec<Value>>,
    pub with_names: Option<bool>,
    pub page_size: Option<i32>,
    pub paging_state: Option<CBytes>,
    pub serial_consistency: Option<Consistency>,
    pub timestamp: Option<i64>,
}
```

Where `query` string could be of two types:

1. `INSERT INTO namespace.table (column_name) VALUE (actual_value)` which contains both column name and actual values that should be inserted

2. `INSERT INTO namespace.table (column_name) VALUE (?)` which is more or less a template that should be used for building query and binding actual values.

`consistency` is a query consistency level.

`values` is an optional list of `Value` (in CDRS version 1.x). Basically `Value` is a `Vec` of bytes and some metadata which helps to identify if a value is `null`, empty or has some value. The order of values needs to be the same as the order of columns in query string. This doesn't allow to use values with names ([one](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L348), [two](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L319))

**Change proposal:** to fix this situation we propose to implement a new enum `QueryValue` that should be used in `Query` params:
```rust
...
pub values: Option<QueryValues>,
...
```
Old `Value` type we're going to keep using to represent protocol's [`[value]`](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L210)

```rust
enum QueryValues {
  SimpleValues(Vec<Value>),
  NamedValues(HashMap<String, Value>)
}
```
 where first option is the same as CDRS has in version 1.x and second one represents new CDRS feature - unordered named values.

There are going to be two implementation of <From<T>> type

```rust
from(value: Vec<Value>) -> QueryValues // SimpleValues
from(value: HashMap<String, Value>) -> QueryValues // NamedValues
```

`Query`'s struct should be changed in order to re-use `ParamsReqQuery`:

```rust
type QueryParams = frame::frame_query::ParamsReqQuery

Query {
  pub query: String,
  pub params: QueryParams
}
```

Query builder should be updated accordingly.

CDRS session should provide methods that would allow more simple usage of queries:

```rust
session.query("INSERT INTO namespace.table (column_name) VALUE (actual_value)");

let with_tracing = true;
let with_warnings = true;
session.exec_tw("INSERT INTO namespace.table (column_name) VALUE (actual_value)", with_tracing, with_warnings);

session.query_with_values("INSERT INTO namespace.table (column_name) VALUE (?)", query_values);
session.query_with_values_tw("INSERT INTO namespace.table (column_name) VALUE (?)", query_values, with_tracing, with_warnings);

session.query_with_params("INSERT INTO namespace.table (column_name) VALUE (actual_value)", query_params);
session.query_with_params("INSERT INTO namespace.table (column_name) VALUE (?)", query_params);
session.query_with_params_tw("INSERT INTO namespace.table (column_name) VALUE (actual_value)", query_params, with_tracing, with_warnings);
session.query_with_params_tw("INSERT INTO namespace.table (column_name) VALUE (?)", query_params, with_tracing, with_warnings);
``` 

Cassandra protocol contains functionality for paging responses. That's why CDRS should provide a handful API for paged response. As an option this API may look like this:

```rust
let pager = session.paged(page_size).query(...);
// and other session exec methods
let rows = paged.next();
```
**Change Proposal**:
* not make request on `paged().exec()`
* make requests on `pager.next()` to get next page
* each value returned by `pager.next()` should be `CDRSResult<Vec<Row>>`
* `pager.is_last()` should return `true` if current page is last and `false` otherwise.

Basically, `session.query` and `pager.query` (`pager.exec_...`) should be considered as different methods.

**Comment by nerdrew** [comment]
Small Question: can the new API use slices instead of Vecs for values?

**Comment by AlexPikalov** [comment]
@nerdrew Sorry for a late replay and thank you for your thoughts on this topic.

Yes, using slices also would be possible if we would use `IntoIterator` as an argument type. It's a good idea actually. Both slice and vector has its implementation, so it seems to be a good solution for this.

**Comment by AlexPikalov** [comment]
UPD. Version 2 development status
- [x] connection to cluster
- [x] password and none authenticator
- [x] lz4, snappy compression 
- [x] cluster query, exec, prepare extended API
- [x] handy paged request API
- [x] `query_values!` macros that allows to easily convert rust values into cassandra values with (for prepared query execution) and without names

Not re-implemented:
- [ ] SSL-encrypted session
- [ ] connection pool
- [ ] server events

Small example https://github.com/AlexPikalov/cdrs/blob/v2.0.0/examples/version_2.rs
